[DOCS-2202] Homere.faivre/synthetics api test timings guide (#11018)
DataDog/documentation: content/en/synthetics/guide/api_test_timing_variations.md

---
title: Understanding API test timings and troubleshooting their variations
kind: documentation
description: Understanding API test timings and troubleshooting their variations
aliases:
- /synthetics/api_test_timing_variations

> further_reading:

Reviewer: Let's maybe add one or two other links (like the two first links in https://github.com/DataDog/documentation/pull/11017/files for isntance)

- link: "https://docs.datadoghq.com/synthetics/metrics/#api-tests"
  tag: "Documentation"

>   text: "Synthetics API test timing metrics"

Reviewer: ```suggestion
  text: "Synthetics API test metrics"
```

---

## Overview

Synthetics API tests collect several [**timing metrics**][1] which can be used to identify any bottlenecks in the communication between the client and your server. 

## Collected Timings

> The metrics collected by Synthetics measure the following timings : 

Reviewer: Add link on metrics

> **1. Redirection time**

Reviewer: What do you think of having ### for all these? And also for the ones in the variations section?

> The `synthetics.http.redirect.time` metric measures the total time spent in redirects. All other network timings (DNS resolution, TCP connection, etc.) correspond to the last request. For example, if an HTTP test (that has the **Follow Redirects** option turned on) first loads page A in 35 ms in total, which redirects to page B that loads in 40ms in total, which redirects to page C, the redirect timing equals 35 + 40 = 75 ms. The load of page C is then split among the various timings (DNS resolution, TCP connection, etc.).

Reviewer: Maybe add link to /synthetics/api_tests/http_tests?tab=requestoptions#define-request for follow redirects

> The `synthetics.http.redirect.time` metric measures the total time spent in redirects. All other network timings (DNS resolution, TCP connection, etc.) correspond to the last request. For example, if an HTTP test (that has the **Follow Redirects** option turned on) first loads page A in 35 ms in total, which redirects to page B that loads in 40ms in total, which redirects to page C, the redirect timing equals 35 + 40 = 75 ms. The load of page C is then split among the various timings (DNS resolution, TCP connection, etc.).

Reviewer: ```suggestion
The `synthetics.http.redirect.time` metric measures the total time spent in redirects. All other network timings (DNS resolution, TCP connection, etc.) correspond to the last request. For example, if an HTTP test (that has the **Follow Redirects** option turned on) first loads page A in 35ms in total, which redirects to page B that loads in 40ms in total, which redirects to page C, the redirect timing equals 35 + 40 = 75ms. The load of page C is then split among the various timings (DNS resolution, TCP connection, etc.).
```

**4. SSL handshake time**

The `synthetics.http.ssl.time` and the `synthetics.ssl.hanshake.time` metrics measure the time spent in SSL handshake. 

These metrics are only collected if the request goes over HTTPS, and not HTTP. 

In case of any redirection, the SSL handshake timing only corresponds to  the last request.

**5. Time to First Byte**

The `synthetics.http.firstbyte.time` metric measures the time between the moment the connection was established and the moment Datadog client received the first byte of the response. Any time spent sending data in the request is included in this timing. 

In case of any redirection, the time to first byte only corresponds to the last request.

**6. Download time**

The `synthetics.http.download.time` metric measures the time between the moment Datadog client receives the first byte of the response and the moment it finishes downloading the whole response. Generally, the bigger the response body, the higher this timing will be. 

In the case the response does not have a body, this timing is null.

In case of any redirections, the download time only corresponds to the last request.

**7. Total response time**

> The `*.response.time`  measures the total time between the moment Synthetics starts the request and the moment it finishes it. The response time is the sum of all network timings. For instance for an HTTP test with no redirections on an HTTPS endpoint, `synthetics.http.response.time = synthetics.http.dns.time + synthetics.http.connect.time+synthetics.http.ssl.time + synthetics.http.donwload.time`.

Reviewer: ```suggestion
The `*.response.time`  measures the total time between the moment Synthetics starts the request and the moment it finishes it. The response time is the sum of all network timings. For instance for an HTTP test with no redirections on an HTTPS endpoint, `synthetics.http.response.time = synthetics.http.dns.time + synthetics.http.connect.time+synthetics.http.ssl.time + synthetics.http.download.time`.
```

## Timing variations
 
Variations in API tests network timing metrics can occur when there is a bottleneck or delay during any of the several stages of the request, from the redirection to the download of the response body. Usually, it is helpful to identify the following : 

- If the variation is observed as a general trend, or as a sudden spike
- If the variation only occurs at specific stage of the request, for instance only for the DNS timings
- If the impacted Synthetics test is running from multiple locations, whether the variation is localised to a single location, or if it is widespread
- If the variation only occurs for a single URL, domain, subdomain, or if it is impacting all tests.

> In many cases, timing can be affected by increase in network latency due to network load

Reviewer: ```suggestion
In many cases, timing can be affected by increase in network latency due to network load.
```

> In many cases, timing can be affected by increase in network latency due to network load

Reviewer: ```suggestion
```

> Common causes by timing

Reviewer: ```suggestion
```

> For each timing metric we measure, variations can usually be explained by several factors.

Reviewer: ```suggestion
For each timing metric we measure, variations can usually be explained by several factors:
```

**Redirection time:** Given this timing is the sum of all redirects in a request, variations at any stage of the HTTP request, from the DNS resolution to the download can noticeably increase the redirection timing. For example any delay in DNS resolution would impact the redirection timing as redirections require API tests to resolve multiple domains.

**DNS resolution time:** An increased DNS resolution time could be due to additional latency from authoritative servers.

> **TCP connection time:** Variations of the TCP handshake can be due to the load of the network and server, the size of the request and response messages, and the distance between the Synthetics managed/private location and the server.

Reviewer: Add link for private location

> **SSL handshake time:** The SSL handshake time can vary depending on the load of the server, as SSL handshakes are usually CPU intensive, but also due to the load of the network and the distance between the Synthetics managed/private location and the server. Issues with CDN also often result in increased SSL handshake time,

Reviewer: ```suggestion
**SSL handshake time:** The SSL handshake time can vary depending on the load of the server, as SSL handshakes are usually CPU intensive, but also due to the load of the network and the distance between the Synthetics managed/private location and the server. Issues with CDN also often result in increased SSL handshake time.
```

Reviewer: Also add link for private location

> **Time To first Byte :** Variations of the Time to First Byte can be due to the network and server load, as well as the distance between the Synthetics managed/private locations and the server. For example a higher network load or the re-routing of traffic caused by an unavailable CDN can negatively impact the Time to First Byte timing.

Reviewer: Add link for private location

**Download time:** Variations in download time can be explained by changes in the size of the response. The downloaded body size is available on test results as well as with the `synthetics.http.response.size` metric.

> For all the above timings where variations can be due to network and server load, it can be useful to leverage [Network Performance monitoring][4] and [Synthetics ICMP tests][5] to identify potential bottlenecks.

Reviewer: How about splitting that sentence into two: 
1. Cases of network load/latency and how users can use NPM, ICMP and TCP tests to identify bottlenecks
2. Cases of server load and how users can use the Agent
?

---

> title: Understanding API test timings and troubleshooting their variations

Reviewer 2: How does "API Test Timing Metrics and Variations" sound as a page title?

```suggestion
title: API Test Timing Metrics and Variations
```

Reviewer 2: Also, how does this page relate to the Synthetic Metrics page (https://docs.datadoghq.com/synthetics/metrics/#metrics)?

Author: IMO, having this page separate will allow us to keep the Synthetics Metrics page short and concise, like a metric 'inventory'. That said, linking the guide in the metrics page will give users possibility to learn more about metrics in the guide.

What do you think @MargotLepizzera

Reviewer: Yeah this one is definitely tricky. 
The first section of the doc would definitely make sense to include in the existing metrics page but it felt like it was too much info to fit in the metrics table @alai97. Plus I believe what's in the table is also used in the `Metrics summary` page and I feel like it would be weird to have so much content in the `Metrics summary`?
The second part however is really about troubleshooting variations so I'd rather keep it separate. 
I think we can go with this for now, and maybe add a card in the docs backlog to think about how to better integrate the first section in the existing metrics page?

Reviewer 2: Gotcha, that rationale sounds good to me!

kind: documentation
description: Understanding API test timings and troubleshooting their variations
aliases:
- /synthetics/api_test_timing_variations
further_reading:
- link: "https://docs.datadoghq.com/synthetics/metrics/#api-tests"
  tag: "Documentation"
  text: "Synthetics API test metrics"
---

## Overview

> Synthetics API tests collect several [**timing metrics**][1] which can be used to identify any bottlenecks in the communication between the client and your server. 

Reviewer 2: Un-bolding the link because "timing metrics" are not UI elements.

```suggestion
You can identify bottlenecks in the communication between your server and the client with [timing metrics][1] collected by Synthetic API tests.

```

> ## Collected Timings

Reviewer 2: ```suggestion
## Collected Timing Metrics
```

> The [metrics][1] collected by Synthetics measure the following timings : 

Reviewer 2: ```suggestion
Synthetic tests collect [metrics][1] that measure the following: 
```

### Redirection time

> The `synthetics.http.redirect.time` metric measures the total time spent in redirects. All other network timings (DNS resolution, TCP connection, etc.) correspond to the last request. For example, if an HTTP test (that has the [**Follow Redirects**][7] option turned on) first loads page A in `35 ms` in total, which redirects to page B that loads in `40 ms` in total, which redirects to page C, the redirect timing equals `35 ms + 40 ms = 75 ms`. The load of page C is then split among the various timings (DNS resolution, TCP connection, etc.).

Reviewer 2: ```suggestion
The `synthetics.http.redirect.time` metric measures the total time spent in redirects. All other network timings (such as DNS resolution and TCP connection) correspond to the last request. 

For example, an HTTP test with **Follow Redirects** selected loads Page A for a total of `35 ms`, which redirects to Page B, which loads for a total of `40 ms`, and redirects to Page C. The redirect timing is calculated as `35 ms + 40 ms = 75 ms` and the load time of Page C is split among all other timings including DNS resolution and TCP connection.

For more information about follow redirects, see [HTTP Tests][7].

```

The `synthetics.http.redirect.time` metric is only measured if redirects occur during the Synthetics HTTP test run. 

### DNS resolution time

> The `*.dns.time` metrics and the `synthetics.dns.response.time` measure the time spent resolving the domain name. Synthetics API tests use common DNS servers for domain name resolution, such as the public Google and CloudFlare DNS as well as some internal AWS and Azure ones. These can be overridden when using [**private locations**][2] or [**DNS tests**][3]. 

Reviewer 2: ```suggestion
The `synthetics.dns.response.time` metric and `*.dns.time` metrics measure the time spent resolving the domain name. Synthetic API tests use common DNS servers for domain name resolution, such as Google, CloudFlare, AWS, and Azure. You can override these servers with [private locations][2] or [DNS tests][3]. 
```

> These metrics are only measured when  the API tests URL field contains a domain name. If it is using an IP address DNS resolution is skipped and there is consequently no time series for these metrics.

Reviewer 2: ```suggestion
These metrics are only measured when the API test URL field contains a domain name. If you use an IP address, DNS resolution is skipped and no time series appear for these metrics.
```

In case of any redirection, the DNS resolution time only corresponds to the last request.

### TCP connection time

The `*.connect.time` metrics measure the total time spent establishing a TCP connection with the server. 

In case of any redirection, the TCP connection time only corresponds to the last request.

### SSL handshake time

The `synthetics.http.ssl.time` and the `synthetics.ssl.hanshake.time` metrics measure the time spent in SSL handshake. 

These metrics are only collected if the request goes over HTTPS, and not HTTP. 

In case of any redirection, the SSL handshake timing only corresponds to  the last request.

> ### Time to First Byte

Reviewer 2: Header titles in sentence case.

```suggestion
### Time to first byte
```

> The `synthetics.http.firstbyte.time` metric measures the time between the moment the connection was established and the moment Datadog client received the first byte of the response. Any time spent sending data in the request is included in this timing. 

Reviewer 2: ```suggestion
The `synthetics.http.firstbyte.time` metric measures the time between the moment the connection was established and the moment the Datadog client received a first byte of the response. This timing includes all time spent sending data in the request. 

```

In case of any redirection, the time to first byte only corresponds to the last request.

### Download time

> The `synthetics.http.download.time` metric measures the time between the moment Datadog client receives the first byte of the response and the moment it finishes downloading the whole response. Generally, the bigger the response body, the higher this timing will be. 

Reviewer 2: ```suggestion
The `synthetics.http.download.time` metric measures the time between the moment the Datadog client receives the first byte of the response and the moment it finishes downloading the entire response. Generally, the bigger the response body, the higher this timing will be. 
```

In the case the response does not have a body, this timing is null.

> In case of any redirections, the download time only corresponds to the last request.

Reviewer 2: Is this word intentionally plural?

```suggestion
In case of any redirection, the download time only corresponds to the last request.
```

### Total response time

> The `*.response.time`  measures the total time between the moment Synthetics starts the request and the moment it finishes it. The response time is the sum of all network timings. For instance for an HTTP test with no redirections on an HTTPS endpoint, `synthetics.http.response.time = synthetics.http.dns.time + synthetics.http.connect.time+synthetics.http.ssl.time + synthetics.http.download.time`.

Reviewer 2: ```suggestion
The `*.response.time` metrics measure the total time between the moment Synthetics starts and the moment Synthetic finishes the request. The response time is the sum of all network timings. 

For example, the total response time for an HTTP test with no redirections on an HTTPS endpoint:  `synthetics.http.response.time = synthetics.http.dns.time + synthetics.http.connect.time+synthetics.http.ssl.time + synthetics.http.download.time`.
```

## Timing variations
 

> Variations in API tests network timing metrics can occur when there is a bottleneck or delay during any of the several stages of the request, from the redirection to the download of the response body. Usually, it is helpful to identify the following : 

Reviewer 2: ```suggestion
Variations in API test network timing metrics can occur when there is a bottleneck or delay in any stage of the request, from the redirection to the download of the response body. 

Identify the following behaviors: 
```

> - If the variation is observed as a general trend, or as a sudden spike

Reviewer 2: ```suggestion
- If the variation is observed as a general trend or a sudden spike
```

> - If the variation only occurs at specific stage of the request, for instance only for the DNS timings

Reviewer 2: ```suggestion
- If the variation only occurs at a specific stage of the request. For example, on the DNS timings.
```

> - If the impacted Synthetics test is running from multiple locations, whether the variation is localised to a single location, or if it is widespread

Reviewer 2: ```suggestion
- If the impacted Synthetics test is running from multiple locations or the variation is localized to a single location or widespread
```

Author: In this case, I believe the original sentence make sense and I don't really agree with the suggestion : the intent here is to ask users to check if the issue is localized or widespread in the case their test is configured on multiple locations.

Reviewer 2: Changed it back to your original sentence structure, just updated the spelling!

> - If the variation only occurs for a single URL, domain, subdomain, or if it is impacting all tests.

Reviewer 2: ```suggestion
- If the variation only occurs for a single URL, domain, or subdomain; or if it is impacting all tests
```

> For each timing metric we measure, variations can usually be explained by several factors:

Reviewer 2: ```suggestion
For every timing metric we measure, variations can be explained with these factors:
```

### Redirection time

> Given this timing is the sum of all redirects in a request, variations at any stage of the HTTP request, from the DNS resolution to the download can noticeably increase the redirection timing. For example any delay in DNS resolution would impact the redirection timing as redirections require API tests to resolve multiple domains.

Reviewer 2: ```suggestion
The redirection time is the sum of all redirects in a request. Variations at any stage of the HTTP request, from the DNS resolution to the download, can noticeably increase the redirection timing. 

For example, any delay in DNS resolution impacts the redirection timing because redirections require API tests to resolve multiple domains.

```

### DNS resolution time

> An increased DNS resolution time could be due to additional latency from authoritative servers.

Reviewer 2: ```suggestion
An increase in DNS resolution time can occur with additional latency from authoritative servers.
```

### TCP connection time

> Variations of the TCP handshake can be due to the load of the network and server, the size of the request and response messages, and the distance between the Synthetics managed/[private location][6] and the server.

Reviewer 2: ```suggestion
Variations of the TCP handshake can occur because of the network and server load, the size of the request and response messages, and the distance between the Synthetics managed and [private location][6] and the server.
```

### SSL handshake time

> The SSL handshake time can vary depending on the load of the server, as SSL handshakes are usually CPU intensive, but also due to the load of the network and the distance between the Synthetics managed/[private location][6] and the server. Issues with CDN also often result in increased SSL handshake time.

Reviewer 2: ```suggestion
Variations of the SSL handshake time can occur because of the server load (SSL handshakes are usually CPU intensive), the network load, and the distance between the Synthetics managed or [private location][6] and the server. Issues with CDN can increase SSL handshake time.
```

> ### Time To first Byte

Reviewer 2: ```suggestion
### Time To first byte
```

> Variations of the Time to First Byte can be due to the network and server load, as well as the distance between the Synthetics managed/[private location][6] and the server. For example a higher network load or the re-routing of traffic caused by an unavailable CDN can negatively impact the Time to First Byte timing.

Reviewer 2: ```suggestion
Variations of the Time to first byte can occur because of the network and server load and the distance between the Synthetics managed or [private location][6] and the server. For example, a higher network load or the rerouting of traffic caused by an unavailable CDN can negatively impact Time to First Byte timing.
```

### Download time

> Variations in download time can be explained by changes in the size of the response. The downloaded body size is available on test results as well as with the `synthetics.http.response.size` metric.

Reviewer 2: ```suggestion
Variations in download time can occur because of changes in the response size. The downloaded body size is available on test results and the `synthetics.http.response.size` metric.
```

> For all above timings where variations can be due to network load, it can be useful to leverage [Network Performance monitoring][4] and [Synthetics ICMP tests][5] to identify potential bottlenecks.

Reviewer 2: ```suggestion
For the above timings where variations can be due to network load, it can be useful to leverage [Network Performance monitoring][4] and [Synthetics ICMP tests][5] to identify potential bottlenecks.

Wherever variations can occur because of network load, you can use [Network Performance Monitoring][4] and [Synthetics ICMP Tests][5] to identify potential bottlenecks.

```

> In cases where variations can be due to server load, leveraging the [Datadog Agent][8] with its various [integrations][9] can help. 

Reviewer 2: ```suggestion

In cases where variations can occur because of server load, use the [Datadog Agent][8] and its [integrations][9]. 
```
